File: gecode/coins3.cpp
/*

  Coin application in Gecode.

 
  From "The ECLiPSe Book" pages 99f and 234 ff
  The solution in ECLiPSe is at page 236.

  """
  What is the minimum number of coins that allows one to pay _exactly_
  any amount smaller than one Euro? Recall that there are six different
  euro cents, of denomination 1, 2, 5, 10, 20, 50
  """

  Compare with the following models:
  * MiniZinc: http://www.hakank.org/minizinc/coins3.mzn
  * Comet: http://www.hakank.org/comet/coins3.co


  This Gecode model was created by Hakan Kjellerstrand (hakank@gmail.com)
  Also, see my Gecode page: http://www.hakank.org/gecode/ .

*/

#include <gecode/driver.hh>
#include <gecode/int.hh>
#include <gecode/minimodel.hh>

using namespace Gecode;

class Coins3 : public MinimizeScript {
protected:

  static const int n = 6; // number of different coins

  int num_coins_val;      // set the number of coins (for showing all solutions)

  IntVarArray x;          // array for number of each coins
  IntVar num_coins;       // number of coins used (to minimize)

public:

  // Search variants
  enum {
    SEARCH_DFS,     // Use depth first search to find the smallest tick
    SEARCH_BAB,     // Use branch and bound to optimize
  };

  Coins3(const SizeOptions& opt) 
  : 
    num_coins_val(opt.size()),
    x(*this, n, 0, 99),
    num_coins(*this, 0, 99)
  {

    // values of the coins
    int _variables[] = {1, 2, 5, 10, 25, 50}; 
    IntArgs variables(n, _variables); 

    // sum the number of coins
    linear(*this, x, IRT_EQ, num_coins, opt.icl());

    // This is the "main loop":
    // Checks that all changes from 1 to 99 can be made
    for(int j = 0; j < 99; j++) {
      IntVarArray tmp(*this, n, 0, 99);
      linear(*this, variables, tmp, IRT_EQ, j, opt.icl());
      for(int i = 0; i < n; i++) {
        rel(*this, tmp[i] <= x[i], opt.icl());
      }
    }

    // set the number of coins (via opt.size())
    // don't forget 
    //  -search dfs
    if (num_coins_val) {
      rel(*this, num_coins == num_coins_val, opt.icl());
    }

    branch(*this, x, INT_VAR_SIZE_MAX(), INT_VAL_MIN()); 

  }

  // Print solution
  virtual void
  print(std::ostream& os) const {
    os << "num_coins: " << num_coins << std::endl;
    os << "x: " << x << std::endl;
    os << std::endl;
  }

  // Constructor for cloning s
  Coins3(bool share, Coins3& s) : MinimizeScript(share,s) {
    x.update(*this, share, s.x);
    num_coins.update(*this, share, s.num_coins);
  }

  // Return cost
  virtual IntVar cost(void) const {
    return num_coins;
  }


  // Copy during cloning
  virtual Space*
  copy(bool share) {
    return new Coins3(share,*this);
  }
};


int
main(int argc, char* argv[]) {
  SizeOptions opt("Coins3");
  opt.solutions(0);

  opt.search(Coins3::SEARCH_BAB);
  opt.search(Coins3::SEARCH_DFS, "dfs");
  opt.search(Coins3::SEARCH_BAB, "bab");

  opt.parse(argc,argv);
  switch (opt.search()) {
    case Coins3::SEARCH_DFS:
      MinimizeScript::run<Coins3,DFS,SizeOptions>(opt); break;
    case Coins3::SEARCH_BAB:
      MinimizeScript::run<Coins3,BAB,SizeOptions>(opt); break;
    }

  return 0;

}

File: sicstus/coins3.pl
/*

  Coin application in SICStus Prolog.
   
  From "The ECLiPSe Book" pages 99f and 234 ff
  The solution in ECLiPSe is at page 236.

  """
  What is the minimum number of coins that allows one to pay _exactly_
  any amount smaller than one Euro? Recall that there are six different
  euro cents, of denomination 1, 2, 5, 10, 20, 50
  """

  Compare with the following model:
  * Comet: http://www.hakank.org/comet/coins3.co


  Model created by Hakan Kjellerstrand, hakank@bonetmail.com
  See also my SICStus Prolog page: http://www.hakank.org/sicstus/

*/

:-use_module(library(clpfd)).
:-use_module(library(lists)).


go :-
       
        % the different coin values
        Variables = [1,2,5,10,25,50],

        length(Variables, N),

        length(Xs,N),
        domain(Xs,0,99),

        % total number of the coins used
        NumCoins in 0..99,
        sum(Xs,#=,NumCoins),
        
        % This is the "main loop":
        % Checks that all changes from 1 to 99 can be made.
        % To be honest I have forgotten exactly how
        % the ECLiPSe solution looked like.
        % Here is one way.
        ( for(J,1,99),
          param(Xs,Variables,N) do
              do_coins(J,Xs,Variables,N)
        ),

        append(Xs,[NumCoins],Vars),
        labeling([minimize(NumCoins)],Vars),

        write(num_coins:NumCoins),nl,
        write(Xs),nl,
        nl,
        fd_statistics.


%
% A related problem:
%
% Is there a set of coin denominations 
% which make it possible to change 
% into 1..99 with less than 8 coins?
% 
% Answer: Yes, there is. For example
% the following configuration:
%
% num_coins:7
% variables:[1,2,3,6,12,25,50]
% x:[1,1,1,1,1,1,1]
% 
% Though, it come with a cost of 
% using 7 different coin denominations.
%
%
% Also, we can see that there is
% a set with just 5 different 
% denominations for the 8 coin 
% change. This is done by changing the
% constraint to: 
%       NumCoins #=< 8,
%
% Result: 
% num_coins:8
% variables:[1,2,6,16,50]
% x:[1,2,2,2,1]
%
go2 :-

        % number of coin denominations
        N in 1..10,
        indomain(N),
        write(n:N),nl,

        % we assume that there are no coin
        % of a denomination larger than 50.
        length(Variables, N),
        domain(Variables,1,50),

        all_distinct(Variables),
        my_ordered(#=<, Variables),

        length(Xs,N),
        domain(Xs,0,8),

        % total number of the coins used
        % in the exchange
        sum(Xs,#=,NumCoins),
        
        % can we manage with less than 8 coins?
        NumCoins #< 8,

        % Note: we cannot use scalar_product/4 here
        % since Variables is not ground any more.
        ( for(J,1,99),
          param(Xs,Variables,N) do
              length(Tmp,N),
              domain(Tmp,0,99),
              ( foreach(X,Xs),
                foreach(T,Tmp),
                foreach(V,Variables),
                fromto(0,In,Out,TmpSum) do
                    T #=< X,
                    Out #= In + V*T
              ),
              TmpSum #= J
        ),

        % search
        append(Xs,Variables,Vars),
        labeling([ffc,bisect,up,minimize(NumCoins)],Vars),

        % output
        write(num_coins:NumCoins),nl,
        write(variables:Variables),nl,
        write(x:Xs),nl,
        nl,
        fd_statistics.



do_coins(J,Xs,Variables,N) :-
        length(Tmp,N),
        domain(Tmp,0,99),
        scalar_product(Variables,Tmp,#=,J),
        ( foreach(X,Xs),
          foreach(T,Tmp) do
              T #=< X
        ).
        

my_ordered(P,List) :-
        ( fromto(List, [This,Next | Rest], [Next|Rest],[_]),
          param(P)
        do
          call(P,This,Next)
        ).
